server/src/main/java/org/elasticsearch/cluster/routing/allocation/Hosts.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.common.Strings;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * {@link Hosts} hosts records the mapping relationship between nodes and hosts,
+ * which makes it easy to find nodes on the same host
+ */
+public class Hosts {
+    private final Map<String, ModelHost> node2Host = new HashMap<>();
+    private final Map<String, ModelHost> hostNameMap = new HashMap<>();
+    private final Map<String, ModelHost> hostAddressMap = new HashMap<>();
```

Comment:
Would be nice to build these maps as local variables in the constructor and then finally assign them as immutable maps to the fields. That would make it clearer that this object is immutable.

Also, the `hostNameMap` and the `hostAddressMap` are only used while constructing the `Hosts` object and thus are better handled as local variables?